File: Eunleelee/2024-02/2월 9일/미로 탈출.py
from collections import deque

def solution(maps):
    def bfs(start, end):
        visit = [[0 for _ in range(R)] for _ in range(C)]

        queue = deque()
        queue.append(start)

        while queue:
            c, r = queue.popleft()

            for i in range(4):
                nc, nr = c + dc[i] , r + dr[i]

                if 0 <= nc < C and 0 <= nr < R and not visit[nc][nr] and maps[nc][nr] != "X":
                    visit[nc][nr] = visit[c][r] + 1
                    queue.append((nc, nr))

        return visit[end[0]][end[1]]

    C = len(maps)
    R = len(maps[0])
    dc = [-1, 1, 0, 0]
    dr = [0, 0, -1, 1]


    for i in range(C):
        for j in range(R):
            if maps[i][j] == "S": start = (i, j)
            elif maps[i][j] == "L": lever = (i, j)
            elif maps[i][j] == "E": end = (i, j)

    distance1 = bfs(start, lever)
    distance2 = bfs(lever, end)

    return distance1 + distance2 if distance1 and distance2 else -1

File: jja6312/2월 9일/미로탈출.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    class Point {
        int x, y, dist;
        boolean leverPulled;

        public Point(int x, int y, int dist, boolean leverPulled) {
            this.x = x;
            this.y = y;
            this.dist = dist;
            this.leverPulled = leverPulled;
        }
    }

    public int solution(String[] maps) {
        int n = maps.length;
        int m = maps[0].length();
        Point start = null, lever = null, exit = null;

        // 지도에서 S, L, E 위치 찾기
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                char c = maps[i].charAt(j);
                if (c == 'S')
                    start = new Point(i, j, 0, false);
                else if (c == 'L')
                    lever = new Point(i, j, 0, false);
                else if (c == 'E')
                    exit = new Point(i, j, 0, false);
            }
        }

        int toLever = bfs(maps, start, lever, n, m);
        if (toLever == -1)
            return -1;

        int toExit = bfs(maps, lever, exit, n, m);
        if (toExit == -1)
            return -1;

        return toLever + toExit;
    }

    private int bfs(String[] maps, Point start, Point end, int n, int m) {
        boolean[][] visited = new boolean[n][m];
        Queue<Point> queue = new LinkedList<>();
        queue.offer(new Point(start.x, start.y, 0, start.leverPulled));

        int[] dx = { 0, 0, 1, -1 };
        int[] dy = { 1, -1, 0, 0 };

        while (!queue.isEmpty()) {
            Point current = queue.poll();
            if (current.x == end.x && current.y == end.y)
                return current.dist;
            for (int i = 0; i < 4; i++) {
                int nx = current.x + dx[i];
                int ny = current.y + dy[i];

                if (nx >= 0 && ny >= 0 && nx < n && ny < m && maps[nx].charAt(ny) != 'X' && !visited[nx][ny]) {
                    visited[nx][ny] = true;
                    queue.offer(new Point(nx, ny, current.dist + 1, current.leverPulled));
                }
            }
        }
        return -1;
    }
}
